sdk/core/Azure.Core/src/Serialization/AzureObjectSerializer.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+
+namespace Azure.Core.Serialization
+{
+    /// <summary>
+    /// An abstraction from reading typed objects.
+    /// </summary>
+    public abstract class AzureObjectSerializer
+    {
+        /// <summary>
+        /// Convert the provided value to it's binary representation and write it to <see cref="System.IO.Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to write to.</param>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        public abstract void Serialize(Stream stream, object? value, Type inputType);
+
+        /// <summary>
+        /// Convert the provided value to it's binary representation and write it to <see cref="System.IO.Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="System.IO.Stream"/> to write to.</param>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="inputType">The type of the <paramref name="value"/> to convert.</param>
+        public abstract ValueTask SerializeAsync(Stream stream, object? value, Type inputType);
```

Comment:
For library authors what's the guidance on when to use sync vs async?

Reply:
If a customer is calling an async method on the client the expectation is that async serialization/deserialization method is invoked and vice-versa.

Reply:
Having two full sync & async implementation is very taxing engineering wise. At-least in Cosmos practically we are overlay sync over async (in Java to start with). 

Do other libraries have full isolated sync and async implementations?

Reply:
Most track 2 HTTP based libraries have 2 isolated code paths for sync and async.